Panic: Non-positive accumulation in reduce

<!--
Thanks for your feedback on Materialize! Please include the following
information in your bug report.
-->

### What version of Materialize are you using?

materialized v0.9.5-dev (1adbb1e7a) 

### How did you install Materialize?

Built from source

### What was the issue?

After stopping, Mz fails to restart:

```
2021-09-16T08:40:33.541589Z ERROR dataflow::render::reduce: [customer-data] Non-positive accumulation in MinsMaxesHierarchical: key: (Row{[]}, 11876133)        value: [Row{[TimestampTz(2021-09-16T08:39:34.287Z)]}, Row{[TimestampTz(2021-09-16T08:39:34.287Z)]}]     count: -1158
2021-09-16T08:40:33.541815Z ERROR dataflow::render::reduce: [customer-data] Non-positive accumulation in MinsMaxesHierarchical: key: (Row{[]}, 11876133)        value: [Row{[TimestampTz(2021-09-16T08:39:34.287Z)]}, Row{[TimestampTz(2021-09-16T08:39:34.287Z)]}]     count: -1158
2021-09-16T08:40:33.543114Z ERROR dataflow::render::reduce: [customer-data] Non-positive accumulation in MinsMaxesHierarchical: key: (Row{[]}, 341797)  value: [Row{[TimestampTz(2021-09-16T08:39:34.287Z)]}, Row{[TimestampTz(2021-09-16T08:39:34.287Z)]}]     count: -1158
2021-09-16T08:40:33.543299Z ERROR dataflow::render::reduce: [customer-data] Non-positive accumulation in MinsMaxesHierarchical: key: (Row{[]}, 14117)   value: [Row{[TimestampTz(2021-09-16T08:39:34.287Z)]}, Row{[TimestampTz(2021-09-16T08:39:34.287Z)]}]     count: -1158
2021-09-16T08:40:33.543444Z ERROR dataflow::render::reduce: [customer-data] Non-positive accumulation in MinsMaxesHierarchical: key: (Row{[]}, 1829)    value: [Row{[TimestampTz(2021-09-16T08:39:34.287Z)]}, Row{[TimestampTz(2021-09-16T08:39:34.287Z)]}]     count: -1158
2021-09-16T08:40:33.543586Z ERROR dataflow::render::reduce: [customer-data] Non-positive accumulation in MinsMaxesHierarchical: key: (Row{[]}, 37)      value: [Row{[TimestampTz(2021-09-16T08:39:34.287Z)]}, Row{[TimestampTz(2021-09-16T08:39:34.287Z)]}]     count: -1158
2021-09-16T08:40:33.543675Z ERROR dataflow::render::reduce: [customer-data] Non-positive accumulation in MinsMaxesHierarchical: key: (Row{[]}, 5)       value: [Row{[TimestampTz(2021-09-16T08:39:34.287Z)]}, Row{[TimestampTz(2021-09-16T08:39:34.287Z)]}]     count: -1158
2021-09-16T08:40:33.543798Z ERROR dataflow::render::reduce: [customer-data] Negative accumulation in ReduceMinsMaxes: [Row{[TimestampTz(2021-09-16T08:39:34.287Z)]}, Row{[TimestampTz(2021-09-16T08:39:34.287Z)]}] with count -1158
2021-09-16T08:40:34.219413Z ERROR panic: called `Option::unwrap()` on a `None` value
```

Stack trace:
```
thread: timely:work-1
location: src/dataflow/src/render/reduce.rs:820:65
version: 0.9.5-dev (1adbb1e7a42564ef7486c3e8f4656908df59c1a9)
backtrace:
   0: materialized::handle_panic
             at src/materialized/src/bin/materialized/main.rs:772:21
   1: core::ops::function::Fn::call
             at /rustc/a178d0322ce20e33eac124758e837cbd80a6f633/library/core/src/ops/function.rs:70:5
   2: std::panicking::rust_panic_with_hook
             at /rustc/a178d0322ce20e33eac124758e837cbd80a6f633/library/std/src/panicking.rs:626:17
   3: std::panicking::begin_panic_handler::{{closure}}
             at /rustc/a178d0322ce20e33eac124758e837cbd80a6f633/library/std/src/panicking.rs:517:13
   4: std::sys_common::backtrace::__rust_end_short_backtrace
             at /rustc/a178d0322ce20e33eac124758e837cbd80a6f633/library/std/src/sys_common/backtrace.rs:141:18
   5: rust_begin_unwind
             at /rustc/a178d0322ce20e33eac124758e837cbd80a6f633/library/std/src/panicking.rs:515:5
   6: core::panicking::panic_fmt
             at /rustc/a178d0322ce20e33eac124758e837cbd80a6f633/library/core/src/panicking.rs:92:14
   7: core::panicking::panic
             at /rustc/a178d0322ce20e33eac124758e837cbd80a6f633/library/core/src/panicking.rs:50:5
   8: dataflow::render::reduce::build_collation::{{closure}}
      differential_dataflow::operators::reduce::ReduceCore::reduce_abelian::{{closure}}
             at /home/moritz/.cargo/git/checkouts/differential-dataflow-d065d23d797aa027/7f1133a/src/operators/reduce.rs:282:21
      <differential_dataflow::operators::reduce::history_replay::HistoryReplayer<V1,V2,T,R1,R2> as differential_dataflow::operators::reduce::PerKeyCompute<V1,V2,T,R1,R2>>::compute
             at /home/moritz/.cargo/git/checkouts/differential-dataflow-d065d23d797aa027/7f1133a/src/operators/reduce.rs:937:29
   9: <differential_dataflow::operators::arrange::arrangement::Arranged<G,T1> as differential_dataflow::operators::reduce::ReduceCore<G,K,V,R>>::reduce_core::{{closure}}::{{closure}}
             at /home/moritz/.cargo/git/checkouts/differential-dataflow-d065d23d797aa027/7f1133a/src/operators/reduce.rs:529:49
  10: <timely::dataflow::stream::Stream<G,D1> as timely::dataflow::operators::generic::operator::Operator<G,D1>>::unary_frontier::{{closure}}::{{closure}}
             at /home/moritz/.cargo/git/checkouts/timely-dataflow-4c0cc365061cd263/4c2886b/timely/src/dataflow/operators/generic/operator.rs:352:17
      timely::dataflow::operators::generic::builder_rc::OperatorBuilder<G>::build::{{closure}}::{{closure}}
             at /home/moritz/.cargo/git/checkouts/timely-dataflow-4c0cc365061cd263/4c2886b/timely/src/dataflow/operators/generic/builder_rc.rs:124:31
      timely::dataflow::operators::generic::builder_rc::OperatorBuilder<G>::build_reschedule::{{closure}}
             at /home/moritz/.cargo/git/checkouts/timely-dataflow-4c0cc365061cd263/4c2886b/timely/src/dataflow/operators/generic/builder_rc.rs:163:26
      <timely::dataflow::operators::generic::builder_raw::OperatorCore<T,L> as timely::scheduling::Schedule>::schedule
             at /home/moritz/.cargo/git/checkouts/timely-dataflow-4c0cc365061cd263/4c2886b/timely/src/dataflow/operators/generic/builder_raw.rs:205:9
  11: timely::progress::subgraph::PerOperatorState<T>::schedule
             at /home/moritz/.cargo/git/checkouts/timely-dataflow-4c0cc365061cd263/4c2886b/timely/src/progress/subgraph.rs:684:30
  12: timely::progress::subgraph::Subgraph<TOuter,TInner>::activate_child
             at /home/moritz/.cargo/git/checkouts/timely-dataflow-4c0cc365061cd263/4c2886b/timely/src/progress/subgraph.rs:344:26
      <timely::progress::subgraph::Subgraph<TOuter,TInner> as timely::scheduling::Schedule>::schedule
             at /home/moritz/.cargo/git/checkouts/timely-dataflow-4c0cc365061cd263/4c2886b/timely/src/progress/subgraph.rs:310:17
  13: timely::progress::subgraph::PerOperatorState<T>::schedule
             at /home/moritz/.cargo/git/checkouts/timely-dataflow-4c0cc365061cd263/4c2886b/timely/src/progress/subgraph.rs:684:30
  14: timely::progress::subgraph::Subgraph<TOuter,TInner>::activate_child
             at /home/moritz/.cargo/git/checkouts/timely-dataflow-4c0cc365061cd263/4c2886b/timely/src/progress/subgraph.rs:344:26
      <timely::progress::subgraph::Subgraph<TOuter,TInner> as timely::scheduling::Schedule>::schedule
             at /home/moritz/.cargo/git/checkouts/timely-dataflow-4c0cc365061cd263/4c2886b/timely/src/progress/subgraph.rs:310:17
  15: timely::worker::Wrapper::step::{{closure}}
             at /home/moritz/.cargo/git/checkouts/timely-dataflow-4c0cc365061cd263/4c2886b/timely/src/worker.rs:749:57
      core::option::Option<T>::map
             at /rustc/a178d0322ce20e33eac124758e837cbd80a6f633/library/core/src/option.rs:489:29
      timely::worker::Wrapper::step
             at /home/moritz/.cargo/git/checkouts/timely-dataflow-4c0cc365061cd263/4c2886b/timely/src/worker.rs:749:26
  16: timely::worker::Worker<A>::step_or_park
             at /home/moritz/.cargo/git/checkouts/timely-dataflow-4c0cc365061cd263/4c2886b/timely/src/worker.rs:396:38
  17: dataflow::server::Worker<A>::run
             at src/dataflow/src/server.rs:587:13
      dataflow::server::serve::{{closure}}
             at src/dataflow/src/server.rs:290:13
      timely::execute::execute::{{closure}}
             at /home/moritz/.cargo/git/checkouts/timely-dataflow-4c0cc365061cd263/4c2886b/timely/src/execute.rs:285:22
  18: timely_communication::initialize::initialize_from::{{closure}}
             at /home/moritz/.cargo/git/checkouts/timely-dataflow-4c0cc365061cd263/4c2886b/communication/src/initialize.rs:291:33
      std::sys_common::backtrace::__rust_begin_short_backtrace
             at /rustc/a178d0322ce20e33eac124758e837cbd80a6f633/library/std/src/sys_common/backtrace.rs:125:18
  19: std::thread::Builder::spawn_unchecked::{{closure}}::{{closure}}
             at /rustc/a178d0322ce20e33eac124758e837cbd80a6f633/library/std/src/thread/mod.rs:481:17
      <std::panic::AssertUnwindSafe<F> as core::ops::function::FnOnce<()>>::call_once
             at /rustc/a178d0322ce20e33eac124758e837cbd80a6f633/library/std/src/panic.rs:347:9
      std::panicking::try::do_call
             at /rustc/a178d0322ce20e33eac124758e837cbd80a6f633/library/std/src/panicking.rs:401:40
      std::panicking::try
             at /rustc/a178d0322ce20e33eac124758e837cbd80a6f633/library/std/src/panicking.rs:365:19
      std::panic::catch_unwind
             at /rustc/a178d0322ce20e33eac124758e837cbd80a6f633/library/std/src/panic.rs:434:14
      std::thread::Builder::spawn_unchecked::{{closure}}
             at /rustc/a178d0322ce20e33eac124758e837cbd80a6f633/library/std/src/thread/mod.rs:480:30
      core::ops::function::FnOnce::call_once{{vtable.shim}}
             at /rustc/a178d0322ce20e33eac124758e837cbd80a6f633/library/core/src/ops/function.rs:227:5
  20: <alloc::boxed::Box<F,A> as core::ops::function::FnOnce<Args>>::call_once
             at /rustc/a178d0322ce20e33eac124758e837cbd80a6f633/library/alloc/src/boxed.rs:1575:9
      <alloc::boxed::Box<F,A> as core::ops::function::FnOnce<Args>>::call_once
             at /rustc/a178d0322ce20e33eac124758e837cbd80a6f633/library/alloc/src/boxed.rs:1575:9
      std::sys::unix::thread::Thread::new::thread_start
             at /rustc/a178d0322ce20e33eac124758e837cbd80a6f633/library/std/src/sys/unix/thread.rs:71:17
  21: start_thread
             at ./nptl/./nptl/pthread_create.c:477:8
  22: __clone
             at ./misc/../sysdeps/unix/sysv/linux/x86_64/clone.S:95
```

### Is the issue reproducible? If so, please provide reproduction instructions.

I can't seem to reproduce the issue, but here's what I did:
1. Clear `mzdata`
2. `cargo run --release --bin materialized -- --workers 16 --introspection-frequency 1s --metrics-scraping-interval 1s --retain-prometheus-metrics 1s`
3. `cargo run --release --bin testdrive -- test/testdrive/prometheus.td`
4. Ctrl+C materialized
5. `cargo run --release --bin materialized -- --workers 4` -> Crashed

### Please attach any applicable log files.

[mzdata.tar.gz](https://github.com/MaterializeInc/materialize/files/7176311/mzdata.tar.gz)

Ah yeah, you're right the +1s and -1s do line up:
```
$ grep '2021-09-16T08:40:06.976Z' repro.stderr | grep -F 'diff=1' | wc -l
     326
$ grep '2021-09-16T08:40:06.976Z' repro.stderr | grep -F 'diff=-1' | wc -l
     326
```

And this agrees with the error message:
```
panic: [customer-data] Non-positive accumulation in MinsMaxesHierarchical: key: (Row{[]}, 165751303)	value: [Row{[TimestampTz(2021-09-16T08:40:06.976Z)]}, Row{[TimestampTz(2021-09-16T08:40:06.976Z)]}]	count: -326
```

All of the +1s are at ts=1631806974083 and the -1s are at ts=1631781906978:
```
$ grep '2021-09-16T08:40:06.976Z' repro.stderr | grep -F 'diff=1' | grep -v '1631806974083'
$ grep '2021-09-16T08:40:06.976Z' repro.stderr | grep -F 'diff=-1' | grep -v '1631781906978'
$
```

--- a/src/persist/src/indexed/mod.rs
+++ b/src/persist/src/indexed/mod.rs
@@ -1108,11 +1108,11 @@ impl Snapshot<Vec<u8>, Vec<u8>> for IndexedSnapshot {
 
 #[cfg(test)]
 mod tests {
-    use std::error::Error;
     use std::sync::mpsc;
 
-    use crate::error::Error as IndexedError;
+    use crate::error::Error;
     use crate::future::Future;
+    use crate::indexed::runtime::DecodedSnapshot;
     use crate::mem::MemRegistry;
 
     use super::*;
@@ -1120,21 +1120,21 @@ mod tests {
     fn block_on_drain<T, F: FnOnce(&mut Indexed<L, B>, FutureHandle<T>), L: Log, B: Blob>(
         index: &mut Indexed<L, B>,
         f: F,
-    ) -> Result<T, IndexedError> {
+    ) -> Result<T, Error> {
         let (tx, rx) = Future::new();
         f(index, tx.into());
         index.drain_pending()?;
         rx.recv()
     }
 
-    fn block_on<T, F: FnOnce(FutureHandle<T>)>(f: F) -> Result<T, IndexedError> {
+    fn block_on<T, F: FnOnce(FutureHandle<T>)>(f: F) -> Result<T, Error> {
         let (tx, rx) = Future::new();
         f(tx.into());
         rx.recv()
     }
 
     #[test]
-    fn single_stream() -> Result<(), Box<dyn Error>> {
+    fn single_stream() -> Result<(), Error> {
         let updates = vec![
             (("1".into(), "".into()), 1, 1),
             (("2".into(), "".into()), 2, 1),
@@ -1231,7 +1231,7 @@ mod tests {
     }
 
     #[test]
-    fn batch_sorting() -> Result<(), Box<dyn Error>> {
+    fn batch_sorting() -> Result<(), Error> {
         let updates = vec![
             (("1".into(), "".into()), 2, 1),
             (("2".into(), "".into()), 1, 1),
@@ -1262,7 +1262,7 @@ mod tests {
     }
 
     #[test]
-    fn batch_consolidation() -> Result<(), Box<dyn Error>> {
+    fn batch_consolidation() -> Result<(), Error> {
         let updates = vec![
             (("1".into(), "".into()), 1, 1),
             (("1".into(), "".into()), 1, 1),
@@ -1301,7 +1301,7 @@ mod tests {
     }
 
     #[test]
-    fn batch_unsealed_empty() -> Result<(), Box<dyn Error>> {
+    fn batch_unsealed_empty() -> Result<(), Error> {
         let mut i = MemRegistry::new().indexed_no_reentrance()?;
         let id = block_on(|res| i.register("0", "", "", res))?;
 
@@ -1327,7 +1327,7 @@ mod tests {
     // Regression test for two similar bugs causing unsealed batches with
     // non-adjacent seqno boundaries (which violates our invariants).
     #[test]
-    fn regression_non_sequential_unsealed_batches() -> Result<(), IndexedError> {
+    fn regression_non_sequential_unsealed_batches() -> Result<(), Error> {
         let mut i = MemRegistry::new().indexed_no_reentrance()?;
 
         // First is some stream is registered, written to, and step'd, moving
@@ -1359,7 +1359,7 @@ mod tests {
     }
 
     #[test]
-    fn test_destroy() -> Result<(), IndexedError> {
+    fn test_destroy() -> Result<(), Error> {
         let mut i = MemRegistry::new().indexed_no_reentrance()?;
 
         let _ = block_on(|res| i.register("stream", "", "", res))?;
@@ -1373,7 +1373,7 @@ mod tests {
         // Destroy stream that was never created.
         assert_eq!(
             block_on(|res| i.destroy("stream2", res)),
-            Err(IndexedError::from(
+            Err(Error::from(
                 "invalid destroy of stream stream2 that was never registered or destroyed"
             ))
         );
@@ -1381,7 +1381,7 @@ mod tests {
         // Creating a previously destroyed stream.
         assert_eq!(
             block_on(|res| i.register("stream", "", "", res)),
-            Err(IndexedError::from(
+            Err(Error::from(
                 "invalid registration: stream stream already destroyed"
             ))
         );
@@ -1390,7 +1390,7 @@ mod tests {
     }
 
     #[test]
-    fn codec_mismatch() -> Result<(), IndexedError> {
+    fn codec_mismatch() -> Result<(), Error> {
         let mut i = MemRegistry::new().indexed_no_reentrance()?;
 
         let _ = block_on(|res| i.register("stream", "key", "val", res))?;
@@ -1401,7 +1401,7 @@ mod tests {
         // Different key codec
         assert_eq!(
             block_on(|res| i.register("stream", "nope", "val", res)),
-            Err(IndexedError::from(
+            Err(Error::from(
                 "invalid registration: key codec mismatch nope vs previous key"
             ))
         );
@@ -1409,10 +1409,48 @@ mod tests {
         // Different val codec
         assert_eq!(
             block_on(|res| i.register("stream", "key", "nope", res)),
-            Err(IndexedError::from(
+            Err(Error::from(
                 "invalid registration: val codec mismatch nope vs previous val"
             ))
         );
+
+        Ok(())
+    }
+
+    #[test]
+    fn regression_8303_snapshot_advance_since() -> Result<(), Error> {
+        let mut i = MemRegistry::new().indexed_no_reentrance()?;
+        let id = block_on(|res| i.register("0", "", "", res))?;
+
+        // Introduce some data, seal it, and advance since. Intentionally don't
+        // call step because might compact it and accidentally produce the right
+        // answer (at the time of the bug, compaction did the right thing, which
+        // is why we didn't catch it initially).
+        let updates = vec![
+            (("1".into(), "".into()), 1, 1),
+            (("1".into(), "".into()), 10, -1),
+            (("2".into(), "".into()), 2, 1),
+        ];
+        block_on_drain(&mut i, |i, res| i.write(vec![(id, updates)], res))?;
+        block_on_drain(&mut i, |i, res| i.seal(vec![id], 4, res))?;
+        block_on_drain(&mut i, |i, res| {
+            i.allow_compaction(vec![(id, Antichain::from_elem(3))], res)
+        })?;
+        let snap = block_on(|res| i.snapshot(id, res))?;
+        // NB: We won't need this hack after the ts advancement logic is moved
+        // to IndexedSnapshot where it should be.
+        let mut snap = DecodedSnapshot::<String, String>::new(snap);
+
+        // Now verify that the snapshot has the right since and that the data in
+        // it has been advanced as expected.
+        assert_eq!(snap.since(), Antichain::from_elem(3));
+        let actual = snap.read_to_end_flattened()?;
+        let expected = vec![
+            (("1".into(), "".into()), 3, 1),
+            (("1".into(), "".into()), 10, -1),
+            (("2".into(), "".into()), 3, 1),
+        ];
+        assert_eq!(actual, expected);
 
         Ok(())
     }

--- a/src/persist/src/indexed/runtime.rs
+++ b/src/persist/src/indexed/runtime.rs
@@ -19,6 +19,7 @@ use std::sync::{Arc, Mutex};
 use std::thread::{self, JoinHandle};
 use std::time::{Duration, Instant};
 
+use differential_dataflow::lattice::Lattice;
 use log;
 use ore::metrics::MetricsRegistry;
 use timely::progress::Antichain;
@@ -536,7 +537,7 @@ pub struct DecodedSnapshot<K, V> {
 }
 
 impl<K: Codec, V: Codec> DecodedSnapshot<K, V> {
-    fn new(snap: IndexedSnapshot) -> Self {
+    pub(crate) fn new(snap: IndexedSnapshot) -> Self {
         DecodedSnapshot {
             snap,
             buf: Vec::new(),
@@ -583,10 +584,23 @@ impl<K: Codec, V: Codec> Snapshot<Result<K, String>, Result<V, String>> for Deco
         &mut self,
         buf: &mut E,
     ) -> Result<bool, Error> {
+        let since = self.since();
         let ret = self.snap.read(&mut self.buf)?;
-        buf.extend(self.buf.drain(..).map(|((k, v), ts, diff)| {
+        buf.extend(self.buf.drain(..).map(|((k, v), mut ts, diff)| {
             let k = K::decode(&k);
             let v = V::decode(&v);
+            // When reading a snapshot, the contract of since is that all update
+            // timestamps will be advanced to it. We do this physically during
+            // compaction, but don't have hard guarantees about how long that
+            // takes, so we have to account for un-advanced batches on reads.
+            //
+            // TODO: This adjustment should logically live in IndexedSnapshot
+            // (where it could be lazy about only doing this on the relevant
+            // batches) but in the short-term it's much easier to do here. I'm
+            // hoping to clean up how snapshots work soon and make them more
+            // like iterators, at which point it will be much simpler to push
+            // this logic where it belongs.
+            ts.advance_by(since.borrow());
             ((k, v), ts, diff)
         }));
         Ok(ret)

--- a/src/persist/src/nemesis/validator.rs
+++ b/src/persist/src/nemesis/validator.rs
@@ -341,19 +341,9 @@ fn updates_eq(
 ) -> bool {
     // TODO: This is also used by the implementation. Write a slower but more
     // obvious impl of consolidation here and use it for validation.
-    //
-    // TODO: The actual snapshot will eventually be compacted up to some since
-    // frontier and the expected snapshot will need to account for that when
-    // checking equality.
 
-    for (_, t, _) in actual.iter_mut() {
-        for ts in since.elements().iter() {
-            if *t < *ts {
-                *t = *ts;
-            }
-        }
-    }
-
+    // The snapshot has been logically compacted to since, so update our
+    // expected to match.
     for (_, t, _) in expected.iter_mut() {
         for ts in since.elements().iter() {
             if *t < *ts {

--- a/src/persist/src/operators/mod.rs
+++ b/src/persist/src/operators/mod.rs
@@ -66,7 +66,15 @@ fn replay<G: Scope<Timestamp = u64>, K: TimelyData + Codec, V: TimelyData + Code
             };
             for update in buf.drain(..) {
                 match flatten_decoded_update(update) {
-                    Ok(u) => ok.push(u),
+                    Ok(u) => {
+                        // The raw update data held internally in the snapshot
+                        // may not be physically compacted up to the logical
+                        // compaction frontier of since. Snapshot handles
+                        // advancing any necessary data but we double check that
+                        // invariant here.
+                        debug_assert!(snapshot.since().less_equal(&u.1));
+                        ok.push(u)
+                    }
                     Err(errs) => errors.extend(errs),
                 }
             }

